SPShellAdmins: Member comparison was case sensitive.

ChristophHannappel opened the pull request:
<!--
    Thanks for submitting a Pull Request (PR) to this project. Your contribution to this project
    is greatly appreciated!

    Please prefix the PR title with the resource name, e.g. 'ResourceName: My short description'.
    If this is a breaking change, then also prefix the PR title with 'BREAKING CHANGE:',
    e.g. 'BREAKING CHANGE: ResourceName: My short description'.

    You may remove this comment block, and the other comment blocks, but please keep the headers
    and the task list.
-->

#### Pull Request (PR) description

The Member comparison of SPShellAdmins was case sensitive. But Active Directory User and Groups are not.
This PR replaced the .Net `.contains` Method with the PowerShell Operator `-contains` and extended the unit tests. 

#### This Pull Request (PR) fixes the following issues

<!--
    If this PR does not fix an open issue, replace this comment block with None. If this PR
    resolves one or more open issues, replace this comment block with a list of the issues using
    a GitHub closing keyword, e.g.:

- Fixes #1439

-->

#### Task list

<!--
    To aid community reviewers in reviewing and merging your PR, please take the time to run
    through the below checklist and make sure your PR has everything updated as required.

    Change to [x] for each task in the task list that applies to your PR. For those task that
    don't apply to you PR, leave those as is.
-->

- [X] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [ ] Resource documentation added/updated in README.md.
- [ ] Resource parameter descriptions added/updated in README.md, schema.mof and comment-based
      help.
- [X] Comment-based help added/updated.
- [ ] Localization strings added/updated in all localization files as appropriate.
- [ ] Examples appropriately added/updated.
- [X] Unit tests added/updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [ ] Integration tests added/updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [X] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

ChristophHannappel commented:
@ykuijs the Build Pipeline has an issue with the GitVersion Task: Calculate ModuleVersion (GitVersion).
I think thats outside of my pr. Could you take a look?

ykuijs commented:
@ChristophHannappel Could you please implement the following change:
Add `  --version 5.12.0` to the end of this line: https://github.com/dsccommunity/SharePointDsc/blob/13408f4f835eeca0e598f093b4f20d1afefd4623/azure-pipelines.yml#L29

The line should say `dotnet tool install --global GitVersion.Tool --version 5.12.0`

ykuijs commented:
An HQRM check is failing. Will check the cause tomorrow.

ChristophHannappel commented:
> An HQRM check is failing. Will check the cause tomorrow.

` Context SharePointDsc.Reverse.psm1
WARNING: The file SharePointDsc.Reverse.psm1 contains a suppression of the required PS Script Analyser rule 
PSAvoidUsingWriteHost. Please remove the rule suppression.
      [-] Should not suppress any required PS Script Analyzer rules 141ms
        Expected $false, but got $true.
        125:                 $requiredRuleIsSuppressed | Should -BeFalse
        at <ScriptBlock>, D:\a\1\s\output\RequiredModules\DscResource.Test\0.16.3\Tests\QA\PSSAResource.common.v4.Tests.ps1: line 125
` 

Thank you :)

johlju commented:
It was a change in module DscResource.Test to make that test run on just the source files. Prior to the change the HQRM test was run on both built module and source files but apparently due to a bug that test was never run on source files before. 🤔 

If you want to keep `Write-Host` I suggest excluding that file for HQRM tests, add the relative path to the file here which I think should ignore the file:

https://github.com/dsccommunity/SharePointDsc/blob/13408f4f835eeca0e598f093b4f20d1afefd4623/build.yaml#L93-L94

Alternative I recommend changing `Write-Host` to:

```powershell
Write-Information -MessageData 'MyMessage' -InformationAction 'Continue'
```

ChristophHannappel commented:
> If you want to keep `Write-Host` I suggest excluding that file for HQRM tests, add the relative path to the file here which I think should ignore the file:
> 
> https://github.com/dsccommunity/SharePointDsc/blob/13408f4f835eeca0e598f093b4f20d1afefd4623/build.yaml#L93-L94

I've added the file to the List, but that did not help.

@ykuijs would it be ok if I replaced the `Write-Host` with `Write-Information` as @johlju suggested?

johlju commented:
Hmm strange I thought that would do it. @ChristophHannappel  Can you test exclude the name `SharePointDsc.Reverse` to verify.

It should filter out the excluded files here:
https://github.com/dsccommunity/DscResource.Test/blob/14acda0bc0a6a9698bfb8fd41da42adfe08b09f0/source/Tests/QA/PSSAResource.common.v4.Tests.ps1#L82

Using the filter: https://github.com/dsccommunity/DscResource.Test/blob/14acda0bc0a6a9698bfb8fd41da42adfe08b09f0/source/Private/WhereSourceFileNotExcluded.ps1#L1

If above doesn't work then there must be a bug in DscResource.Test, need to debug to find it. I submit a bug in that repo.  Maybe @dan-hughes have some input?

But do suggest replace `Write-Host` so hopefully @ykuijs is okay with that.

johlju commented:
> It was a change in module DscResource.Test to make that test run on just the source files. Prior to the change the HQRM test was run on both built module and source files but apparently due to a bug that test was never run on source files before. 🤔

Not true, that change hasn't been released yet. My bad. There must be something else going on in 0.16.3 release.

@ChristophHannappel can we test to see if the latest preview will help with this, there are a lot of improvements there:

https://github.com/dsccommunity/SharePointDsc/blob/13408f4f835eeca0e598f093b4f20d1afefd4623/RequiredModules.psd1#L20

too:

``` powershell
    'DscResource.Test'     = @{
        Version    = 'latest'
        Parameters = @{
            AllowPrerelease = $true
        }
    }
```` 

That will use the preview release 0.17.0-preview0003.

ykuijs commented:
@ChristophHannappel, not sure what the impact will be if we remove Write-Host. I know we have been using Write-Host to specifically write to the screen instead of in the pipeline so we can use formatting, etc.

I see that the tests keep failing. In the last change you didn't specify the file extension. Can you please add that?

johlju commented:
> not sure what the impact will be if we remove Write-Host. I know we have been using Write-Host to specifically write to the screen instead of in the pipeline so we can use formatting, etc.

The command `Write-Information` uses the information stream to output messages to the console. But it cannot use text color without using ansi-sequences, but PS5 doesn't handle ansi-sequences either so text color is not possible. More input: https://stackoverflow.com/questions/38523369/write-host-vs-write-information-in-powershell-5

johlju commented:
@ChristophHannappel thanks for testing the changes, it seems it didn't do anything. I won't have time to debubg DscResource.Test, but could try to do it this weekend - it could probably be a problem in the future for others as well. 

@ykuijs  I really suggest to move away from Write-Host for this particular problem, then the code will still be tested correctly and not excluded from all HQRM tests.

dan-hughes commented:
HQRM issue is probably [this](https://github.com/dsccommunity/DscResource.Test/commit/ccfded1c7947384beb5eabcc60a38bc7653b2ca2) commit.

I do not see any further changes to the Pester v4 HQRM tests.

johlju commented:
I added this issue https://github.com/dsccommunity/DscResource.Test/issues/142 to track that it is not possible to exclude a source file from HQRM testing.

dan-hughes commented:
dsccommunity/DscResource.Test#143 waiting to be merged that fixes the HQRM issue you are having.

The path in `build.yml` required is the path to the psm1 file relative to the source directory which in this case is the SharePointDsc directory inside the module.

ChristophHannappel commented:
> @ChristophHannappel Once the other PR is merged: Can you update the build.yaml with the string suggested by @dan-hughes, so that the correct path will be excluded?

Sure thing.

@dan-hughes: will that also be a preview release?

dan-hughes commented:
Yes. Unless one of the maintainers decides to trigger a full release.

johlju commented:
We can test this PR with the preview and if it works I can release a full release of DscResource.Test. But I wont have time to review @dan-hughes PR until this weekend. To much to do at the day job 😊

dan-hughes commented:
I'd like to take credit, but the PR was pulled as it wasn't the correct fix.

Nice to know what the workaround is though!

ChristophHannappel commented:
Seems like I have to fix some tests. 
@ykuijs do you have a tip how to run the tests only for one Resource? So I can avoid the whole build.ps1 experience. Thank you

dan-hughes commented:
> Seems like I have to fix some tests. @ykuijs do you have a tip how to run the tests only for one Resource? So I can avoid the whole build.ps1 experience. Thank you

You can just specify the path to the test file in the Script key in `build.yml`.

It will not help with code coverage but it helps with test runtime.

johlju commented:
> t will not help with code coverage but it helps with test runtime.

If the code coverage is extensive it can be turned of by also passing `-CodeCoverageThreshold 0`

It should also possible to use `Invoke-Pester` directly after running `./build.ps1 -Task noop` (onse for each session to set up the environment).

ykuijs commented:
I believe it is also possible to run the `.\build.ps1 -Tasks Build` once and then just run the ps1 file of the Pester test.

ykuijs commented:
@ChristophHannappel Now that all tests have completed successfully, can this PR be merged?

ChristophHannappel commented:
@ykuijs Yes please.

At Everyone: Thank you for helping so quickly with my pipeline issues
